mcp/src/main/java/io/modelcontextprotocol/server/McpAsyncServerExchange.java

```diff
@@ -126,7 +128,16 @@ public Mono<McpSchema.ElicitResult> createElicitation(McpSchema.ElicitRequest el
 	 * @return A Mono that emits the list of roots result.
 	 */
 	public Mono<McpSchema.ListRootsResult> listRoots() {
-		return this.listRoots(null);
+
+		return this.listRoots(McpSchema.FIRST_PAGE).expand(result -> {
+			if (result.nextCursor() != null) {
+				return this.listRoots(result.nextCursor());
+			}
+			return Mono.empty();
+		}).reduce(new McpSchema.ListRootsResult(new ArrayList<>(), null), (allRootssResult, result) -> {
+			allRootssResult.roots().addAll(result.roots());
+			return allRootssResult;
```

Comment:
The backing list needs to be made unmodifiable after being accumulated.

Reply:
I made some adjustments , but not sure if this is what you've meant.